sockeye/optimizers.py

```diff
@@ -0,0 +1,202 @@
+    :param beta4: Exponential decay rate for checkpoint objective relative change.
+    :param epsilon: Small value to avoid division by 0.
+    :param k_lo: Lower threshold for relative change.
+    :param k_hi: Upper threshold for relative change.
+    :param use_batch_objective: Incorporate batch objective (can use both).
+    :param use_checkpoint_objective: Incorporate checkpoint objective (can use both).
+    """
+    def __init__(self,
+                 learning_rate: float = 0.001,
+                 beta1: float = 0.9,
+                 beta2: float = 0.999,
+                 beta3: float = 0.999,
+                 beta4: float = 0.,
+                 epsilon: float = 1e-8,
+                 k_lo: float = 0.1,
+                 k_hi: float = 10,
+                 use_batch_objective: bool = True,
+                 use_checkpoint_objective: bool = False,
+                 **kwargs):
+        check_condition(any((use_batch_objective, use_checkpoint_objective)),
+                        "Must use at least one of: batch objective, checkpoint objective")
+        super().__init__(learning_rate=learning_rate, **kwargs)
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+        self.beta4 = beta4
+        self.epsilon = epsilon
+        self.k_lo = k_lo
+        self.k_hi = k_hi
+        self.use_batch_objective = use_batch_objective
+        self.use_checkpoint_objective = use_checkpoint_objective
+
+    def create_state(self, index: int, weight: NDArray):
+        return (zeros(weight.shape, weight.context, dtype=weight.dtype),  # mean
+                zeros(weight.shape, weight.context, dtype=weight.dtype),  # variance
+                [0, 1],     # batch previous objective "hat", previous d
+                [0, 0, 1])  # checkpoint number, previous objective "hat", previous d
+
+    def update(self,
+               index: int,
+               weight: NDArray,
+               grad: NDArray,
+               state: object):
+
+        assert isinstance(weight, NDArray)
+        assert isinstance(grad, NDArray)
+        lr = self._get_lr(index)
+
+        mean, var, prev_batch, prev_checkpoint = state
+
+        # Standard Adam rules for updating mean and variance
+        self._update_count(index)
+        t = self._index_update_count[index]
+
+        coef1 = 1. - self.beta1**t
+        coef2 = 1. - self.beta2**t
+        lr *= math.sqrt(coef2)/coef1
+
+        mean[:] = self.beta1 * mean + (1. - self.beta1) * grad
+        var[:] = self.beta2 * var + (1. - self.beta2) * (grad**2)
+
+        # Now compute Eve's d term
+        d = 0.
+
+        # Eve rules from paper: compute f_hat and d based on last training batch objective
+        if self.use_batch_objective:
+            f_hat_prev, d_prev = prev_batch
+            f = self.batch_state.metric_val
+            if t > 1:
+                # The original paper has a typo in the algorithm here.  The following lines are re-
```

Comment:
computation and update of these intermediate values is duplicated for batch and checkpoint objective. Could you consider factoring this out into a common method?

Reply:
That's probably a better way to do things for now.  Some of the extensions we've been looking at can treat batch/checkpoint values differently so these may not always be identical.